### PR TITLE
added substring, json, len, is_valid methods

### LINKED ISSUE #722: Implement existing operations in Python client
As a new iteration of Kaskada written in Python, I want to add functional and aggregation parity with the existing Kaskada service so that I can enable a transition story from existing users of Kaskada.

Acceptance Criteria:
* Determine the subset of functions required for implementation from the existing [Function Catalog ](https://kaskada.io/docs-site/kaskada/main/fenl/catalog.html). These include but are not limited to:

- [x] lower
- [x] round
- [x] sqrt
- [x] upper
- [x] zip_max (greatest)
- [x] zip_min (least)
- [x] hash
- [x] substring
- [x] len

* There are additional time functions that need to be considered as well, but could be separated out as an individual ticket. See #729

### PR DESCRIPTION
**Pull request by epinzur**
this closes #722

### CODE REVIEW

File: python/docs/source/reference/timestream/misc.md [hunk 1]
```diff
@@ -12,6 +12,7 @@
     Timestream.filter
     Timestream.hash
     Timestream.if_
+    Timestream.is_valid
```
Comment (bjchambers) [inline python/docs/source/reference/timestream/misc.md:15]: `is_valid` was added as `is_not_null` in `logical.md`. I don't think we need to add it.

File: python/docs/source/reference/timestream/string.md [hunk 2]
```diff
@@ -6,6 +6,9 @@
 .. autosummary::
    :toctree: ../apidocs/
 
+    Timestream.json
```
Comment (bjchambers) [inline python/docs/source/reference/timestream/string.md:9]: If we can't get `json` to work as it was, we can either drop it for now, or expose the primitive `json_field` which takes the field name: `e.json_field("name")`. The support was previously relying on detecting `json(e).name` and converting it to that instruction, but that is only part of the Fenl compilation path, not what the Python stuff goes through.

I think maybe we should skip JSON for now.

File: python/pytests/substring_test.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            '1996-12-19T16:40:01,B,,2,"salutations"',
+            '1996-12-19T16:40:02,B,"goodbye",,',
+        ]
+    )
+    return await kd.sources.CsvString.create(
+        content, time_column="time", key_column="key"
+    )
+
+
+async def test_substring(source, golden) -> None:
```
Comment (bjchambers) [inline python/pytests/substring_test.py:23]: We may also want to test with negative numbers (I think substring uses that to reflect "distance from the end"). So `start=-1` would capture the last character, etc.

File: python/pysrc/kaskada/sources/arrow.py [hunk 4]
```diff
@@ -246,6 +246,10 @@ def __init__(
             strings_can_be_null=True,
         )
 
+    _parse_options = pyarrow.csv.ParseOptions(
+        escape_char="\\",
+    )
+
```
Comment (epinzur) [inline python/pysrc/kaskada/sources/arrow.py:252]: this was done to allow escaping comma's in CSV to handle JSON.  should I remove for now since we aren't adding json anymore?

File: python/pytests/substring_test.py (hunk 3 above)
Comment (epinzur) [reply python/pytests/substring_test.py:23]: this already exists via line 12, 31, 32

File: python/docs/source/reference/timestream/string.md (hunk 2 above)
Comment (epinzur) [reply python/docs/source/reference/timestream/string.md]: removed. issue made to do later #768

File: python/pysrc/kaskada/_timestream.py [hunk 5]
```diff
@@ -321,6 +321,14 @@ def hash(self) -> Timestream:
         """
         return Timestream._call("hash", self)
 
+    def len(self) -> Timestream:
+        """Return a Timestream with the length of input string.
```
Comment (bjchambers) [inline python/pysrc/kaskada/_timestream.py:325]: May want this to support lists as well. Not sure if we need to do anything *now*.

File: python/pysrc/kaskada/sources/arrow.py (hunk 4 above)
Comment (bjchambers) [reply python/pysrc/kaskada/sources/arrow.py:252]: seems OK to leave. fine either way.

**Review by bjchambers — APPROVED** [review]
